src/compiler/scala/tools/nsc/transform/AccessorSynthesis.scala

```diff
@@ -349,50 +300,52 @@ trait AccessorSynthesis extends Transform with ast.TreeDSL {
-            debuglog("adding checked getter for: " + getter + " " + lhs.symbol.flagString)
-            List(typedPos(pos)(mkSetFlag(getter)))
-          }
-          else Nil
-        }
         override def transformStats(stats: List[Tree], exprOwner: Symbol) = {
-          // !!! Ident(self) is never referenced, is it supposed to be confirming
-          // that self is anything in particular?
-          super.transformStats(
-            stats flatMap {
-              case stat@Assign(lhs@Select(This(_), _), rhs) => stat :: checkedGetter(lhs)(stat.pos.focus)
-              // remove initialization for default values -- TODO is this case ever hit? constructors does not generate Assigns with EmptyTree for the rhs AFAICT
-              case Apply(lhs@Select(Ident(self), _), EmptyTree.asList) if lhs.symbol.isSetter => Nil
-              case stat                                                                       => List(stat)
-            },
-            exprOwner
-          )
+          val checkedStats = stats flatMap {
+            // Mark field as initialized after an assignment
+            case stat@Assign(lhs@Select(This(_), _), _) =>
+              stat :: bitmapOf(lhs.symbol).toList.map(bitmap => typedPos(stat.pos.focus)(mkSetFlag(bitmap)))
+
+            // remove initialization for default values
+            // TODO is this case ever hit? constructors does not generate Assigns with EmptyTree for the rhs AFAICT
+            // !!! Ident(self) is never referenced, is it supposed to be confirming
+            // that self is anything in particular?
+            case Apply(lhs@Select(Ident(self), _), EmptyTree.asList) if lhs.symbol.isSetter => Nil
+            case stat => List(stat)
+          }
+
+          super.transformStats(checkedStats, exprOwner)
         }
       }
 
+      private[this] val isTrait = clazz.isTrait
+      // We only act on concrete methods, and traits only need to have their constructor rewritten
+      def needsWrapping(dd: DefDef) =
+        dd.rhs != EmptyTree && (!isTrait || dd.symbol.isConstructor)
+
       /** Make getters check the initialized bit, and the class constructor & setters are changed to set the initialized bits. */
-      def wrapRhsWithInitChecks(sym: Symbol)(rhs: Tree): Tree = {
-        // Add statements to the body of a constructor to set the 'init' bit for each field initialized in the constructor
+      def wrapRhsWithInitChecks(sym: Symbol)(rhs: Tree): Tree =
         if (sym.isConstructor) addInitBitsTransformer transform rhs
-        else if (isTrait || rhs == EmptyTree) rhs
-        else if (needsInitFlag(sym)) // getter
-          mkCheckedAccessorRhs(if (isUnitGetter(sym)) UNIT else rhs, rhs.pos, sym)
-        else if (sym.isSetter) {
-          val getter = sym.getterIn(clazz)
-          if (needsInitFlag(getter)) Block(List(rhs, typedPos(rhs.pos.focus)(mkSetFlag(getter))), UNIT)
-          else rhs
+        else if ((sym hasFlag ACCESSOR) && !(sym hasFlag LAZY)) {
```

Comment:
This should also have excluded `PARAMACCESSOR` --> `&& !(sym hasFlag LAZY | PARAMACCESSOR)`